geode-core/src/main/java/org/apache/geode/cache/query/internal/index/IndexManager.java

```diff
@@ -1418,6 +1431,9 @@ public String getCanonicalizedIteratorName(String definition) {
     }
 
     public void addTask(int action, RegionEntry entry, int opCode) {
+      if (isDone()) {
```

Comment:
when we only set the status to true when there is no other tasks in the queue?

Reply:
isDone()'s javadoc says it's for testing, and is a synchronized call, I am not sure if adding the check here would have some synchronization issues. Seems like more synchronous tests need to be added.

Reply:
When first task is added to queue, is when index maintenance is started. I agree that some more test are needed for this part.

Reply:
I agree with @jinmeiliao here, the method is only used by tests and it's clearly not intended to be used as part of production code:

```
    /**
     * Used by tests to determine if the updater thread has finished updating its indexes. The list
     * is cleared without synchronization, which makes this methods somewhat unsafe from a threading
     * point of view.
     */
    public synchronized boolean isDone() {
      return this.pendingTasks.size() == 0;
    }

```

Reply:
This setting/unsetting is going to have a lot of race conditions where after we have checked the flag, it will skip over and miss being under maintenance.

For example Thread 1 hit the isDone() method, it comes back as false.  We don't set the boolean in the index.  T1 is context switched out and whatever is in the pending task list is completed.  A GFSH user checks  the index and it says DONE.  T1 data has not updated the index.   T1 awakens and continues to add a pending task but the boolean is still false.

Reply:
Thanks for comments. I will update solution.